### ISSUE TITLE
Layout structure

### ISSUE THREAD

**Issue by phproberto** [joomla-projects/cms-naked#2]

**Comment by Bakual** [comment]
Trying to put together the current facts (correct me if I'm wrong)

### Current folders and useage

/layouts/joomla  -> _contains basically all layouts used by core extensions_
/layouts/libraries  -> _contains layouts for JHtmlBootstrap tabs_

### Current override structure for components

```
[0] => templates/mytemplate/html/layouts/com_mycomponent
[1] => components/com_mycomponent/layouts
[2] => templates/mytemplate/html/layouts
[3] => layouts
```

#### Thus locations for extension layouts for best override support are:

/components/com_example/layouts
/plugins/content/example/layouts
/modules/mod_example/layouts

### Useful docs (done by Roberto)

http://docs.joomla.org/Sharing_layouts_across_views_or_extensions_with_JLayout
http://magazine.joomla.org/issues/issue-nov-2013/item/1590-jlayout-layouts-improvements-joomla-3-2

**Comment by Bakual** [comment]
JLayouts allows to customise it so you can basically load layouts from different extensions by passing in parameters. As an extension developer you can technically load layouts from any place you want, by specifying a different extension, client (site or backend) and / or by adding custom includepaths.

The point we need to find an answer is where do we want to store the layouts in future so we can choose the path for that today.

One option is to keep everything in the extension folder. More or less like we do today. If you search the output file of a given extension, you'd have to search in the path of that extension.

Another option is to move all output to a centralised folder. Like what is currently the case with the existing JLayouts.

#### Personal view:

The second option has the advantage that we can build the exact same paths for the default layouts and the template overrides. A template designer could just copy-paste the whole folder into his template and start adjusting the files as he likes.
With the current way, there are some folders which are in reversed order which is a bit unintuitive till you figured it out. It's not bad with JLayouts itself, but with older override structures like view layouts and media files (css, js, ...) this is really funny to figure out.

It also follows the way how JLayouts are handled by core currently.

Another advantage is that we would truly decouple presentation from logic. Presentation would all be in the /layouts folder.

The biggest disadvantage I see is that the /layouts folder will get cluttered by all the extensions putting their layouts there. This would need some clear guidance how that folder structure should look. Like for example use `/layouts/com_example/` as base.
This could even be enforced if the installer would automatically put the layouts in the correct folder.

Another disadvantage could be for the extension developer. The files would be more spread across the site.

**Comment by anibalsanchez** [comment]
Previous analysis are complete. 

I think it's better to have /layouts/com_example/.

If not, it's arbitrary to have /layouts folder at the CMS root for "core". At the end com_content, could have its own layouts in com_content/layouts. You could remove /layouts folder and distribute layouts, and avoid all the fuss of a new root folder.

**Comment by mbabker** [comment]
How do you handle layouts which are supposed to be reusable regardless of
the component using it then?

**Comment by phproberto** [comment]
As not everybody is in the skype discussion this is more or less what has been discussed. Feedback is welcome before approve it.

The main plan is to create a default template in:

`templates/default`

That will be the core template. That template cannot be activated / removed because it's something to use internaly and as reference.

Layouts will be moved to:

`templates/default/layouts`

We will create a layout installer without allowing developer to select the layouts destination folder. Layouts will be automatically installed in:

`templates/default/layouts`

To create a custom override you will copy the desired layout into the active template layouts folder and then override it:

`templates/mytemplate/layouts`

### Component layouts

Components will integrate the support for new views based on layouts. That will allow us to render views wherever we need them (HMVC).

Main layouts/views folder will be:

`templates/default/layouts/components/com_name`

Views will be inside:

`templates/default/layouts/components/com_name/views/viewname/default.php`

Overrides will be inside:

`templates/mytemplate/layouts/components/com_name`

### Libraries layouts

Layouts will be inside:

`templates/default/layouts/libraries/name`

Overrides will be inside:

`templates/mytemplate/layouts/libraries/name`

### Modules layouts

Layouts will be inside:

`templates/default/layouts/modules/name`

Overrides will be inside:

`templates/mytemplate/layouts/modules/name`

### Plugins layouts

Layouts will be inside:

`templates/default/layouts/plugins/type/name`

Overrides will be inside:

`templates/mytemplate/layouts/plugins/type/name`

**Comment by anibalsanchez** [comment]
Perfect. Core components and third-party extensions at the same level of re-use and overrides.

**Comment by betweenbrain** [comment]
Excellent!

I like the idea of having everything in one place, and grouped like this, but  what is the advantage of creating a default template for this? Why not just create a default layout directory like `/layouts/default`, `layouts/default/components/com_name`... etc.? Thanks for explaining that,  I may be missing something.

**Comment by Fedik** [comment]
if this allow keep all markup in same place (especially after View can use Layout), it songs logical :)

but small question, after that we will have `templates/default` and `templates/system` that a bit confuse. Maybe we can merge it somehow?

**Comment by phproberto** [comment]
@betweenbrain it has some benefits over the current `layouts` folder:
- **Designers have a template to use reference to override things**. No fear to break anything. Copy&paste, modify and done.
- **Backenders / frontenders roles are better identified**. Everything inside templates is frontend stuff. Note that backenders still provide the default / basic layout.
- Easier to maintain. **Rename any template to default makes it the new base template**. So for example in v4 we can replace all the layouts base markup. All the logic tied to the layout loading will still work.

@Fedik I agree. That's something we have to deal with. Always ensuring BC.

**Comment by betweenbrain** [comment]
Thanks @phproberto!

A couple of thoughts and questions to explore some ideas:
- I do understand about providing a base template to work from, and the fear of breaking something, but I do think there will be to much confusion with having `templates/default` and `templates/system`. With that in mind, why do we even need to have `templates/system` exposed as we don't people modifying it (a.k.a. it is core files)? I'd suggest considering eliminating `templates/system` in favor of moving those views further into core.
- Why do we need to separate front and backend? In fact, I think it is a very cool idea of having all layouts, front and backend, in one place so that they can be reused everywhere.

> Rename any template to default makes it the new base template

Nice!

From a code perspective, would the `JLayoutHelper::render();` calls change? I suspect not.

**Comment by phproberto** [comment]
@betweenbrain  thanks for your feedback.

> I do think there will be to much confusion with having templates/default and templates/system. With that in mind, why do we even need to have templates/system exposed as we don't people modifying it (a.k.a. it is core files)? I'd suggest considering eliminating templates/system in favor of moving those views further into core.

As I said to @Fedik I agree we need to remove `templates/system`. I think that's something easy to do once we have this decided. 

> Why do we need to separate front and backend? In fact, I think it is a very cool idea of having all layouts, front and backend, in one place so that they can be reused everywhere.

If that comes from my previous frontend/backend explanation probably I haven't explained it right. I'm talking about the frontenders vs backenders roles, not about frontend & backend. Sorry for the confusion.

Purely talking about frontend / backend I think that's too much for this. Let's do things right to make everything fully overridable and with a good layout structure. In the future we can mix frontend & backend. Now I think it can confuse us all. We still don't even have layouts :D

> From a code perspective, would the JLayoutHelper::render(); calls change? I suspect not.

No, all has to be BC. 

The main idea is to use a function `getRenderer()` from each main extension type. This is the sample I proposed for views:

``` php
/**
 * Get the renderer
 *
 * @param   string  $layout   Layout to load.
 * @param   array   $options  Optional array with layout options
 *
 * @return  JLayoutFile       Renderer instance
 */
public function getRenderer($layout, $options = array())
{
    $component = JApplicationHelper::getComponentName();
    $template  = JFactory::getApplication()->getTemplate();

    $renderer = new JLayoutFile($layout, null, $options);

    $renderer->setIncludePaths(
        array(
            JPATH_THEMES . '/' . $template . '/layouts/components/' . $component . '/views/' . $this->getName(),
            JPATH_THEMES . '/default/layouts/components/' . $component . '/views/' . $this->getName()
        )
    );

    return $renderer;
}
```

so you can use it like:

``` php
echo  $this->getRenderer('default')->render(
    array(
        'items'      => $this->items,
        'pagination' => $this->pagination,
        'state'      => $this->state,
        'user'       => $this->user
    )
);
```

That's the code that would be placed inside controllers' `display` method.

**Comment by betweenbrain** [comment]
Thanks @phproberto, I like the direction where this is going.

> If that comes from my previous frontend/backend explanation probably I haven't explained it right. I'm talking about the frontenders vs backenders roles, not about frontend & backend. Sorry for the confusion.

I'm likely missing something more fundamental to understand the reason for this change, and I apologize for seeming a bit confused about the reason for this change. Are the layouts that we have today in `/layouts` only usable in the frontend? Is the change to emphasize frontenders/template devs using layouts more?

**Comment by babsgosgens** [comment]
Deferring the responsibility of rendering a view to the template will be a tremendous improvement to our current workflow. Client side developers will have much more control over their output, and the server side devs will no longer need to think about how to present their output (creating views). All they need to do is prepare whatever data needs to be passed to the view. Then, rendering that data will be the responsibility of a template.

Love that.

I do have my reservations about the proposed file paths though. For one thing, having views in a subfolder to layouts makes little sense to me. If anything, it should be the other way around, considering a view will likely be composed of one or more layouts.

But, there is another thing that has me thinking: layouts are something of a higher hierarchy. Layouts may be re-used over different templates, and do (not necessarily) need to be part of any specific template. I kinda look at them as a library of html snippets. To have to copy them inside different templates goes against our desire to keep code DRY.

If the general consensus is to keep them in the templates folder, then something along these lines makes a little more sense to me:

`templates/layouts/namespace/layout.php`

For example:
`templates/layouts/pagination/default.php`
or
`templates/layouts/lists/ordered.php`

As we are already using namespaces in our plugins, it will not take much effort for devs to get on board with that.

Component layouts would than just be layouts in an existing namespace, or in a custom one.

Now, layouts can be distributed to templates that request them.

However, if this makes better sense for some, then this may also be a valid approach:
`templates/default/layouts/namespace/layout.php`

In either case, overrides would be in:
`templates/mytemplate/layouts/namespace/layout.php`

**Views**
Now to views. Views in a template are essentially what is now the html folder. So views would be placed in a context in the same way that we are now storing layout overrides:

`templates/mytemplate/views/components/com_name/viewname/default.php`

The way I see it, views would essentially be replacing the html folder.

Admittedly, this scenario introduces the problem of linking menu items up to the correct view - but anything can be solved right ;).

So, these are my 2cts, looking forward to your views.

**Comment by Bakual** [comment]
> I'm likely missing something more fundamental to understand the reason for this change, and I apologize for seeming a bit confused about the reason for this change. Are the layouts that we have today in /layouts only usable in the frontend? Is the change to emphasize frontenders/template devs using layouts more?

@betweenbrain You don't miss something and the question is valid. `/layouts` would work perfectly fine from anywhere.
The reason for this proposed paths is not really a technical one. It's more for making it easier for templaters. If we can put _all_ output into a single folder like `/templates/default`, it would become dead simple to create a new template or add new overrides. Today you need to know that to override `/layouts/foo/bar.php` you need to create `templates/my_template/html/layouts/foo/bar.php`. That's halfway intuitive for JLayouts but as you know it's more complicate for views and even worse for media files. Sometimes you have to leave out a folder (like `tmpl`) or need to switch the order of the folders to a proper override.
If we can put everything into `/templates/default` instead, we could have the exact same folder structure there as you would need for your overrides. You could just copy the whole folder and rename it and you got your own template. Or to create overrides for a new extension you just copy paste that extension folder to your template.

**Comment by Bakual** [comment]
> Client side developers will have much more control over their output, and the server side devs will no longer need to think about how to present their output (creating views). All they need to do is prepare whatever data needs to be passed to the view.

That's not 100% correct. Extension developers would still need to provide default outputs for their stuff like today. But it would be stored in `/template/default` somewhere. If they don't do that, their extension wouldn't work in case the template doesn't have the overrides already.
In fact it doesn't change that much technically. It would just make it clearer that the output is a template thing. And this hopefully raises the awareness for extension developers. And it would be much more intuitive and clear how to override that output in your own template.

**Comment by betweenbrain** [comment]
Thanks @Bakual, that makes perfect sense. Knowing that the goal is to make it as easy as possible to override the core does give this a different context in my mind. Very interesting... I'm intrigued!

**Comment by anibalsanchez** [comment]
It also enables an extension installer to place layouts in a known folder, shared with the rest of exensions.

**Comment by babsgosgens** [comment]
@Bakual you're right of course - extension developers would still create some sort of output - layouts as well as views.

Would still like to see layouts separate from the template though. With that I mean as not part of an actual template. They can still be in the 'templates' folder - or in the root, either approach works for me.

**Comment by Fedik** [comment]
about `templates/system`, maybe we can do like next: 
- tell Joomla! use `default` as default template instead of `system`
- move `error.php`, `offline.php`, `component.php` to `templates/default` - so we keep good examples for user
- move `css` and `images` somewhere in `media` 
- use current files in `system` as proxy to files that in `default` for BC (example in `component.php` add `include default/component.php`)
- and in Joomla! 4 we just delete `system` :smirk:

about `modChrome`, maybe we can replace this old stuff to `layouts`?
something like `patch.to.chrome.none`, `patch.to.chrome.html5` and so on... only question where the right place to store it? :wink:

**Comment by Bakual** [comment]
Good idea to do the module chromes with JLayouts. It may be a bit tricky due to the way we load them but should be doable I think. It would certainly be more intuitive having them as layouts in `/templates/my_template/layouts/chrome` or so.
